### PR TITLE
internal/charmstore: rework download statistics

### PR DESCRIPTION
**Pull request by mhilton**

### CODE REVIEW

File: internal/charmstore/store.go [hunk 1]
```diff
@@ -1422,22 +1419,27 @@ func (s StoreDatabase) Migrations() *mgo.Collection {
 	return s.C("migrations")
 }
 
+// Macaroons returnst the Mongo collection where macaroon info is stored.
```
Comment (ajbeattie) [inline internal/charmstore/store.go]: s/returnst/returns

File: internal/charmstore/store.go [hunk 2]
```diff
@@ -1422,22 +1419,27 @@ func (s StoreDatabase) Migrations() *mgo.Collection {
...
 func (s StoreDatabase) Macaroons() *mgo.Collection {
 	return s.C("macaroons")
 }
 
+// DownloadCounts returnst the Mongo collection where download counts are stored.
```
Comment (ajbeattie) [inline internal/charmstore/store.go]: s/returnst/returns

File: internal/storetesting/stats/stats.go [hunk 3]
```diff
@@ -12,43 +12,91 @@ import (
...
 			}
 			return
 		}
 	}
-	c.Errorf("total downloads for %#v is %d, want %d", id, doc.TotalDownloads, expected)
+	c.Errorf("total downloads for %#v is %d, want %d", id, counts.Total, expected)
+}
+
+// ThisWeek processes the given offsets downloads offsets and counts, and
+// calculates how many occured in the current week. This is necessary as
```
Comment (ajbeattie) [inline internal/storetesting/stats/stats.go]: s/occured/occurred

File: internal/storetesting/stats/stats.go [hunk 4]
```diff
@@ -12,43 +12,91 @@ import (
...
+// weekly stats are grouped by ISO8601 week, and therefore the value
+// changes with depending on how far through the week you are.
```
Comment (ajbeattie) [inline internal/storetesting/stats/stats.go]: s/with//

File: internal/storetesting/stats/stats.go [hunk 5]
```diff
@@ -12,43 +12,91 @@ import (
...
+			if i <= days {
+				total += int64(n)
+			}
+		}
+	}
+
+	return total
+}
+
+//TthisMonth processes the given offsets downloads offsets and counts, and
```
Comment (ajbeattie) [inline internal/storetesting/stats/stats.go]: Space at beginning of doc and s/TthisMonth/ThisMonth/

File: internal/storetesting/stats/stats.go [hunk 6]
```diff
@@ -12,43 +12,91 @@ import (
...
+// calculates how many occured in the  current month. This is necessary as
```
Comment (ajbeattie) [inline internal/storetesting/stats/stats.go]: s/occured/occurred

File: internal/storetesting/stats/stats.go [hunk 7]
```diff
@@ -12,43 +12,91 @@ import (
...
+// monthly stats are grouped by month, and therefore the value
+// changes with depending on how far through the month you are.
```
Comment (ajbeattie) [inline internal/storetesting/stats/stats.go]: s/with//

File: internal/storetesting/stats/stats.go [hunk 8]
```diff
@@ -12,43 +12,91 @@ import (
...
 				continue // Wait a bit to make sure.
 			}
 			return
 		}
 	}
-	c.Errorf("total downloads for %#v is %d, want %d", id, doc.TotalDownloads, expected)
+	c.Errorf("total downloads for %#v is %d, want %d", id, counts.Total, expected)
+}
+
+// ThisWeek processes the given offsets downloads offsets and counts, and
```
Comment (ajbeattie) [inline internal/storetesting/stats/stats.go]: s/offsets downloads offsets/download offsets/  ? As mentioned in comment below, mentioning how the offsets work would be helpful I think.

File: internal/storetesting/stats/stats.go [hunk 9]
```diff
@@ -12,43 +12,91 @@ import (
...
+// calculates how many occured in the current week. This is necessary as
+// weekly stats are grouped by ISO8601 week, and therefore the value
+// changes with depending on how far through the week you are.
+func ThisWeek(args ...map[int]int) int64 {
```
Comment (ajbeattie) [inline internal/storetesting/stats/stats.go:60]: Could you add a comment (or add to godoc) expanding a bit more on the data structure of the args, here and for `ThisMonth` below? The backwards-offset is not obvious here (it's possible more context is available elsewhere outside of this PR, but I think it'd be helpful to include it here too).

**Review by ajbeattie — APPROVED** [review]
LGTM with a series of typo/format nitpicks and a comment request.

File: internal/charmstore/stats.go [hunk 10]
```diff
@@ -340,16 +100,129 @@ func (s *Store) IncrementDownloadCountsAtTime(id *router.ResolvedURL, t time.Tim
...
 	}
 	return nil
 }
+
+func (s *Store) incrementDownloadCountsAtTime(url *charm.URL, t time.Time) error {
+	day, dayExpires := currentDay(t)
+	week, weekExpires := currentWeek(t)
+	month, monthExpires := currentMonth(t)
+
+	id := url.String()
```
Comment (tasdomas) [inline internal/charmstore/stats.go]: The `withRevision` / `withoutRevision` naming used earlier was a tiny bit clearer.

File: internal/v5/stats.go [hunk 11]
```diff
@@ -44,73 +42,7 @@ func parseDateRange(form url.Values) (start, stop time.Time, err error) {
...
-			Count: entry.Count,
-		}
-		if !entry.Time.IsZero() {
-			stat.Date = entry.Time.Format("2006-01-02")
-		}
-		items = append(items, stat)
-	}
-
-	return items, nil
+	return nil, errgo.WithCausef(nil, params.ErrForbidden, "cannot retrieve stats")
```
Comment (alesstimec) [inline internal/v5/stats.go:45]: nitpick: maybe a 410 would be a more appropriate code..

**Review by alesstimec — APPROVED** [review]
LGTM

**Review by tasdomas — APPROVED** [review]

**Pull request merged**